Implement Vulkan Presentation Engine + Rework GraphicBufferProducer & VI

PixelyIon opened the pull request:
The following items were performed in this PR:
* Implement [Vulkan](https://www.vulkan.org/) [`Instance`](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkInstance.html) + [`PhysicalDevice`](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPhysicalDevice.html) + [`Device`](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkDevice.html) initialization with [`VK_EXT_debug_report`](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VK_EXT_debug_report.html) support
* Integrate [`VK_LAYER_KHRONOS_validation`](https://github.com/KhronosGroup/Vulkan-ValidationLayers/blob/master/docs/khronos_validation_layer.md) from Android NDK into debug builds
* Implement base Vulkan infrastructure (`FenceCycle`, Memory Manager [using [VMA](https://github.com/GPUOpen-LibrariesAndSDKs/VulkanMemoryAllocator)] and Command Scheduler)
* Rework `GuestTexture` + `Texture` classes with Vulkan infrastructure
* Implement Vulkan-based Presentation Engine
* Isolate guest and host presentation for allowing [n-Buffering](https://en.wikipedia.org/wiki/Multiple_buffering) on host
* Utilize [`AChoreographer`](https://developer.android.com/ndk/reference/group/choreographer) for accurate V-Sync events and adaptive refresh rate updates
* Dynamic swap-interval management using [`VK_GOOGLE_display_timing`](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VK_GOOGLE_display_timing.html)
* Implement cropping, transformation and scaling using host [Surfaceflinger](https://source.android.com/devices/graphics/surfaceflinger-windowmanager) for [HWC](https://source.android.com/devices/graphics/hwc) acceleration
* Rework Performance Statistics to provide a more accurate representation of performance
* Introduce `Display` settings for fine-grained control over presentation
* Rework `GraphicBufferProducer` for accuracy and (re)implement the following transactions:
* * `RequestBuffer`
* * `SetBufferCount`
* * `DequeueBuffer`
* * `DetachBuffer`
* * `DetachNextBuffer`
* * `AttachBuffer`
* * `QueueBuffer`
* * `CancelBuffer`
* * `Query`
* * `Connect`
* * `Disconnect`
* * `SetPreallocatedBuffer`
* Rework Host1X Syncpoint for code quality and performance
* Fix race condition in Host1X Syncpoints affecting titles like [Sonic Mania](https://www.nintendo.com/games/detail/sonic-mania-switch/)
* Rework [Display Services (`VI`)](https://switchbrew.org/wiki/Display_services) + [`IHOSBinderDriver`](https://switchbrew.org/wiki/Nvnflinger_services#dispdrv) for accuracy
* Implement SVC [`UnmapSharedMemory`](https://switchbrew.org/wiki/SVC#UnmapSharedMemory)

Closes #197 
Closes #198

PixelyIon committed:
Vulkan Instance + Validation Layer + Debug Report Initialization

The GPU class has been extended in this for Vulkan initialization, this is done to the point of initializing the instance alongside loading in `VK_LAYER_KHRONOS_validation` which is also now packed into all Debug APKs for Skyline. In addition, `VK_EXT_debug_report` is also initialized and it's output is piped directly into the Logger. 

A minor change regarding the type of the `Fps` and `Frametime` globals was changed to `skyline::i32`s which is a more suitable type due to those having a smaller chance of overflowing while being signed as Java doesn't have unsigned integral types.

PixelyIon committed:
Vulkan PhysicalDevice + Device Initialization + Report Accurate Version

Vulkan Device initialization is handled now, it supports required extensions but support for optional extensions/features/properties will come in later when we require those. In addition, we now correctly report the version of Skyline to Vulkan which can be accessed from debugging tools. 

There's also a minor change regarding the search pattern for `SkylineLibraries` which now only searches in headers of libraries and it also explicitly excludes the redundant `vulkan.hpp` from the `Vulkan-Headers` repository.

PixelyIon committed:
Extend VkQueue Allocation Code

The VkQueue allocation code is rewritten to iterate through all queue families to look for an appropriate queue now

PixelyIon committed:
Update NDK to 21.4.7075529

PixelyIon committed:
Presentation Engine Vulkan Surface + Swapchain Initialization

This commit adds in VkSurface/VkSwapchain initialization and recreation. It also adapts GraphicsBuffferProducer and Texture to fit in with those changes but it doesn't yet implement presenting those buffers nor uploading guest buffers onto the host.

PixelyIon committed:
Rework GraphicBufferProducer

This commit makes GraphicBufferProducer significantly more accurate by matching the behavior of AOSP alongside mirroring the tweaks made by Nintendo. 

It eliminates a lot of the magic structures and enumerations used prior and replaces them with the correct values from AOSP or HOS. 

There was a lot of functional inaccuracy as well which was fixed, we emulate the exact subset of HOS behavior that we need to. A lot of the intermediate layers such as GraphicBufferConsumer or Gralloc/Sync are not emulated as they're pointless abstractions here.

PixelyIon committed:
Implement Fence Cycle, Memory Manager and Command Scheduler

Implements a wrapper over fences to track a single cycle of activation, implement a Vulkan memory manager that wraps the Vulkan-Memory-Allocator library and a command scheduler for scheduling Vulkan command buffers

PixelyIon committed:
Redesign Texture Class + Improve Presentation Engine

This commit reworks the `Texture` class to include a Vulkan Image backing that can be optionally owning or non-owning and swapped in with consideration for Vulkan image layout, it also adds CPU-sided synchronization for the texture objects with FenceCycle. It also makes the appropriate changes to `PresentationEngine` and `GraphicBufferProducer` to work with the new `Texture` class while setting the groundwork for supporting swapchain recreation. It also fixes a log in `IpcResponse` and improves the display mode selection algorithm by further weighing refresh rate.

PixelyIon committed:
Separate Guest and Host Presentation + AChoreographer V-Sync Event

We had issues when combining host and guest presentation since certain configurations in guest presentation such as double buffering were very unoptimal for the host and would significantly affect the FPS. As a result of this, we've now made host presentation have its own presentation textures which are copied into from the guest at presentation time, allowing us to change parameters of the host presentation independently of the guest.

We've implemented the infrastructure for this which includes being able to create images from host GPU memory using VMA, an optimized linear texture sync and a method to do on-GPU texture-to-texture copies.

We've also moved to driving the V-Sync event using AChoreographer on its on thread in this PR, which more accurately encapsulates HOS behavior and allows games such as ARMS to boot as they depend on the V-Sync event being signalled even when the game isn't presenting.

PixelyIon committed:
Revamp Host1X Syncpoint + Address Review Comments

PixelyIon committed:
Implement Display Settings + Attach Texture to FenceCycle

PixelyIon committed:
Replace LogActivity with "Share Logs" + Add Timestamp to Logger

PixelyIon committed:
Update C++ Libraries

PixelyIon committed:
Support Swap Interval + Cropping + Transforms and more for QueueBuffer

Support for the following parameters was added to `QueueBuffer`:
* Earliest Present Timestamp
* Swap Interval
* Crop
* Scaling Mode
* Transform
* Frame ID (Not returned to guest yet)
We utilize ANativeWindow APIs directly to achieve all of this in an efficient manner since HWC will be used directly for it, we do plan to introduce Vulkan equivalents for all of these operations later down the line for a port to non-Android platforms.

PixelyIon committed:
Rework Performance Statistics

We used instantaneous values for FPS previously which led to a lot of variation in it and the inability to determine a proper FPS value due to constant fluctuations. All FPS values are now averaged to allow reading out a stable value and a deviation statistic has been added for the frame-time to judge judder and frame-pacing which allows for a significantly better measure of overall performance. The formatting for all the floating-point numbers is now fixed-point to prevent shifting of position due to decimal digits becoming 0.

PixelyIon committed:
Improve `ENUM_STRING` + Fix Host1X Syncpoints

`ENUM_STRING` now has a unified implementation in <common/macros.h> with a documented format and can be used throughout the codebase. 

A major performance regression was added in the Host1X Syncpoint revamp as it did a syscall if there were any waiters during `Increment` even if they would just be woken up and go back to sleep as the threshold wasn't hit. It has now been optimized to only do a wake if any waiting thread needs to be awoken. 

There was also a bug concerning increment where it would perform actions corresponding to the previous increment rather than the current one which has also been fixed.

PixelyIon committed:
Rework VI + IHOSBinder

VI/IHOSBinder suffered from major inaccuracies in their function due to being quickly thrown together initially with little concern for accuracy, this has now been fixed with them being substantially more accurate now.

PixelyIon committed:
Implement SvcUnmapSharedMemory + Remove Redundant [[unlikely]]

PixelyIon committed:
Implement additional IGraphicBufferProducer transactions

The following GraphicBufferProducer transactions were implemented:
* `SetBufferCount`
* `DetachBuffer`
* `DetachNextBuffer`
* `AttachBuffer`
It should be noted that `preallocatedBufferCount` (previously `hasBufferCount`) and `activeSlotCount` were adapted accordingly with how they were effectively the same value as all active buffers were preallocated prior but now there can be a non-preallocated active slot. 
Additionally, a bug has been fixed where `SetPreallocatedBuffer` has the graphic buffer as an optional argument whereas it was treated as a mandatory argument prior and could lead to a SEGFAULT if an application were to not pass in a buffer.

PixelyIon committed:
Support Sticky Transforms + Minor Formatting Fixes

Sticky transforms have been stubbed, as they are on HOS/Android. Certain titles like Xenoblade Chronicles end up setting the sticky transform even if it doesn't do anything, as a result of this we cannot throw an exception for it and stub it without an exception (Aside from the cases where the value isn't recognized).

bylaws commented on app/src/main/cpp/skyline/gpu/texture/texture.h:
```diff
@@ -0,0 +1,352 @@
+        Texture(GPU &gpu, texture::Dimensions dimensions, const texture::Format &format, vk::ImageLayout initialLayout = vk::ImageLayout::eGeneral, vk::ImageUsageFlags usage = {}, vk::ImageTiling tiling = vk::ImageTiling::eOptimal, vk::ComponentMapping mapping = {});
+
+        /**
+         * @brief Acquires an exclusive lock on the texture for the calling thread
```
Note that these are lower case so they can be used with unique_lock

PixelyIon replied:
See the class comment `@note This class conforms to the Lockable and BasicLockable C++ named requirements`

bylaws commented on app/src/main/cpp/skyline/kernel/memory.cpp:
```diff
@@ -62,7 +62,7 @@ namespace skyline::kernel {
 
         auto result{mmap(reinterpret_cast<void *>(base.address), base.size, PROT_NONE, MAP_FIXED | MAP_ANONYMOUS | MAP_PRIVATE, -1, 0)};
-        if (result == MAP_FAILED) [[unlikely]]
+        if (result == MAP_FAILED)
```
I don't understand why you removed this?

PixelyIon replied:
It's implicit for any paths that throw an exception, the compiler automatically understands that they're unlikely branches

bylaws commented on app/src/main/cpp/skyline/kernel/types/KSharedMemory.cpp:
```diff
@@ -43,15 +43,33 @@ namespace skyline::kernel::type {
+        if (guest.ptr != ptr && guest.size != size)
+            throw exception("Unmapping KSharedMemory partially is not supported: Requested Unmap: 0x{:X} - 0x{:X} (0x{:X}), Current Mapping: 0x{:X} - 0x{:X} (0x{:X})", ptr, ptr + size, size, guest.ptr, guest.ptr + guest.size, guest.size);
+
+        if (mmap(ptr, size, PROT_NONE, MAP_ANONYMOUS | MAP_PRIVATE | MAP_FIXED, 0, 0) == MAP_FAILED)
```
no unlikely here? I don't understand our policy on this

PixelyIon replied:
Same as above, our policy is effectively `[[unlikely]]` on any unlikely path that isn't implicitly marked as such

bylaws commented on app/src/main/cpp/skyline/services/hosbinder/GraphicBufferProducer.cpp:
```diff
@@ -1,214 +1,702 @@
-    Buffer::Buffer(const GbpBuffer &gbpBuffer, std::shared_ptr<gpu::Texture> texture) : gbpBuffer(gbpBuffer), texture(std::move(texture)) {}
+    GraphicBufferProducer::GraphicBufferProducer(const DeviceState &state) : state(state), bufferEvent(std::make_shared<kernel::type::KEvent>(state, true)) {}
+
+    u8 GraphicBufferProducer::GetPendingBufferCount() {
```
Any reason for a `u8` here when they're more expensive

bylaws commented on app/src/main/cpp/skyline/services/hosbinder/GraphicBufferProducer.cpp:
```diff
@@ -1,214 +1,702 @@
+    AndroidStatus GraphicBufferProducer::DetachNextBuffer(std::optional<GraphicBuffer> &graphicBuffer, std::optional<AndroidFence> &fence) {
+        std::scoped_lock lock(mutex);
+        auto bufferSlot{queue.end()};
+        for (auto it{queue.begin()}; it != queue.end(); it++) {
```
Could find_if work here?

PixelyIon replied:
See how it doesn't `break` after finding one, it needs to find the oldest `Free` slot rather than just any `Free` slot in the queue

bylaws replied:
So reversed find_if?

PixelyIon replied:
What if the first slot is the oldest? The queue isn't ordered by age

bylaws commented on app/src/main/cpp/skyline/services/hosbinder/GraphicBufferProducer.cpp:
```diff
@@ -1,214 +1,702 @@
+
+        auto &handle{graphicBuffer.graphicHandle};
+        if (handle.magic != NvGraphicHandle::Magic)
+            throw exception("Unexpected NvGraphicHandle magic: {}", handle.surfaceCount);
```
I too have a surfacecount magic

PixelyIon replied:
In my defense, a lot of copy pasting was involved :stuck_out_tongue:

bylaws commented on app/src/main/cpp/skyline/services/hosbinder/GraphicBufferProducer.cpp:
```diff
@@ -1,214 +1,702 @@
+            throw exception("Multi-planar surfaces are not supported: {}", handle.surfaceCount);
+
+        auto &surface{graphicBuffer.graphicHandle.surfaces.at(0)};
+        if (surface.scanFormat != NvDisplayScanFormat::Progressive)
```
540i interlaced CRT skyline :(

PixelyIon replied:
I'll let you get Khronos Group to support it in Vulkan first then I can add support for interlaced output to Skyline

bylaws commented on app/src/main/cpp/skyline/services/hosbinder/GraphicBufferProducer.cpp:
```diff
@@ -1,214 +1,702 @@
+        slot = std::distance(queue.begin(), bufferSlot);
+
+        preallocatedBufferCount = std::count_if(queue.begin(), queue.end(), [](const BufferSlot &slot) { return slot.graphicBuffer && slot.isPreallocated; });
+        activeSlotCount = std::count_if(queue.begin(), queue.end(), [](const BufferSlot &slot) { return static_cast<bool>(slot.graphicBuffer); });
```
```suggestion
        activeSlotCount = std::count_if(queue.begin(), queue.end(), [](const BufferSlot &slot) { return slot.graphicBuffer != nullptr; });
```

bylaws commented on app/src/main/cpp/skyline/services/hosbinder/GraphicBufferProducer.cpp:
```diff
@@ -1,214 +1,702 @@
+        preallocatedBufferCount = std::count_if(queue.begin(), queue.end(), [](const BufferSlot &slot) { return slot.graphicBuffer && slot.isPreallocated; });
+        activeSlotCount = std::count_if(queue.begin(), queue.end(), [](const BufferSlot &slot) { return static_cast<bool>(slot.graphicBuffer); });
+
+        state.logger->Debug("#{} - Dimensions: {}x{} [Stride: {}], Format: {}, Layout: {}, {}: {}, Usage: 0x{:X}, NvMap {}: {}, Buffer Start/End: 0x{:X} -> 0x{:X}", slot, surface.width, surface.height, handle.stride, ToString(graphicBuffer.format), ToString(surface.layout), surface.layout == NvSurfaceLayout::Blocklinear ? "Block Height" : "Pitch", surface.layout == NvSurfaceLayout::Blocklinear ?  ...
```
wrap this line

bylaws commented on app/src/main/cpp/skyline/services/hosbinder/GraphicBufferProducer.cpp:
```diff
@@ -1,214 +1,702 @@
+                tileMode = gpu::texture::TileMode::Block;
+                tileConfig = {
+                    .surfaceWidth = static_cast<u16>(surface.width),
+                    .blockHeight = static_cast<u8>(1U << surface.blockHeightLog2),
```
blockHeightBits is a much better varname for surface.blockHeightLog2

PixelyIon replied:
The units involved aren't bits though, height isn't measured in bits and I think `log2` is consistent with the TRM

bylaws commented on app/src/main/cpp/skyline/services/hosbinder/GraphicBufferProducer.cpp:
```diff
@@ -1,214 +1,702 @@
+            if (graphicBuffer->magic != GraphicBuffer::Magic)
+                throw exception("Unexpected GraphicBuffer magic: 0x{} (Expected: 0x{})", graphicBuffer->magic, GraphicBuffer::Magic);
+            else if (graphicBuffer->intCount != sizeof(NvGraphicHandle) / sizeof(u32))
+                throw exception("Unexpected GraphicBuffer native_handle integer count: 0x{} (Expected: 0x{})", graphicBuffer->intCount, sizeof(NvGraphicHandle));
```
```suggestion
                throw exception("Unexpected GraphicBuffer native_handle integer count: 0x{} (Expected: 0x{})", graphicBuffer->intCount, sizeof(NvGraphicHandle) / sizeof(u32));
```

bylaws commented on app/src/main/cpp/skyline/services/hosbinder/GraphicBufferProducer.cpp:
```diff
@@ -1,214 +1,702 @@
+        }
+
+        preallocatedBufferCount = std::count_if(queue.begin(), queue.end(), [](const BufferSlot &slot) { return slot.graphicBuffer && slot.isPreallocated; });
+        activeSlotCount = std::count_if(queue.begin(), queue.end(), [](const BufferSlot &slot) { return static_cast<bool>(slot.graphicBuffer); });
```
```suggestion
        activeSlotCount = std::count_if(queue.begin(), queue.end(), [](const BufferSlot &slot) { return slot.graphicBuffer != nullptr; });
```

bylaws commented on app/src/main/cpp/skyline/services/hosbinder/GraphicBufferProducer.cpp:
```diff
@@ -1,214 +1,702 @@
+            }
+
+            case TransactionCode::Connect: {
+                auto hasProducerListener{static_cast<bool>(in.Pop<u32>())};
```
```suggestion
                bool hasProducerListener{in.Pop<u32>() != 0};
```

bylaws commented on app/src/main/cpp/skyline/services/hosbinder/GraphicBufferProducer.h:
```diff
@@ -1,160 +1,191 @@
+        std::mutex mutex; //!< Synchronizes access to the buffer queue
+        constexpr static u8 MaxSlotCount{16}; //!< The maximum amount of buffer slots that a buffer queue can hold, Android supports 64 but they go unused for applications like games so we've lowered this to 16 (https://cs.android.com/android/platform/superproject/+/android-5.1.1_r38:frameworks/native/include/gui/BufferQueueDefs.h;l=29)
+        std::array<BufferSlot, MaxSlotCount> queue;
+        u8 activeSlotCount{}; //!< The amount of slots in the queue that can be dequeued
```
use an i32 for these things, android does so it makes sense to

PixelyIon replied:
I use `unsigned` as a guarantee here, the code will break very hard if any of these values are negative

bylaws replied:
well im sure android's code would too? if not use u32

PixelyIon replied:
There's no need to make the type wider either since `MaxSlotCount` is `u8` and this is guaranteed to never go above that, all this would do is make the loads/stores go from `LDRB`/`STRB` to `LDR`/`STR` which are marginally less efficient but for no reason, none of the arithmetic on the CPU side uses this directly so no instructions will be generated from the usage of it here.

bylaws replied:
but android itself supports more slots than a u8 no?

PixelyIon replied:
It actually doesn't, Android just likes to be more inefficient I suppose?

bylaws replied:
Word

PixelyIon replied:
> none of the arithmetic on the CPU side uses this directly so no instructions will be generated from the usage of it here.
Anyway, this is sort of pedantic so we'll address it if it comes up in the profiler till then we can take advantage of the lower memory footprint :P

bylaws commented on app/src/main/cpp/skyline/services/hosbinder/IHOSBinderDriver.cpp:
```diff
@@ -7,43 +7,166 @@
+            throw exception("Adjusting Binder object reference count for unknown object: #{}", binderHandle);
+
+        auto value{request.Pop<i32>()};
+        bool isStrong{static_cast<bool>(request.Pop<u32>())};
```
```suggestion
        bool isStrong{request.Pop<u32>() != 0};
```

bylaws commented on app/src/main/cpp/skyline/services/hosbinder/IHOSBinderDriver.h:
```diff
@@ -3,17 +3,46 @@
       private:
-        std::shared_ptr<GraphicBufferProducer> producer;
+        DisplayId displayId{DisplayId::Null}; //!< The ID of the display that the layer is connected to
+        constexpr static u64 DefaultLayerId{1}; //!< The VI ID of the default (and only) layer in our surface stack
```
can you not mix the constants like this, add some spacing or smth

bylaws commented on app/src/main/cpp/skyline/services/hosbinder/parcel.h:
```diff
@@ -67,12 +103,9 @@ namespace skyline::service {
-            }
+            auto offset{objects.size()};
+            objects.resize(offset + sizeof(ObjectType));
+            *(reinterpret_cast<ObjectType *>(objects.data() + offset)) = object;
```
I think you may as well use memcpy here? like it's a bit more understandable

bylaws commented on app/src/main/java/emu/skyline/MainActivity.kt:
```diff
@@ -120,15 +122,25 @@ class MainActivity : AppCompatActivity() {
 
         binding.searchBar.apply {
-            binding.logIcon.setOnClickListener { startActivity(Intent(context, LogActivity::class.java)) }
+            binding.logIcon.setOnClickListener {
```
I think this would look better if it were written in ☕ 🧻

PixelyIon replied:
Two people disagree so you're clearly wrong here, it looks the best in Kotlin :sunglasses:

PixelyIon commented on app/build.gradle:
```diff
@@ -15,7 +15,7 @@ android {
 
         versionCode 3
-        versionName "0.3"
+        versionName "0.0.3"
```
I did this since we need a 3 component version for Vulkan, I did make it `0.0.3` rather than `0.3.0` by mistake but does this matter? It's sort of abstract anyway but I could change it to that if you really want it to be `0.3.0`

bylaws replied:
personally prefer 1.0.0 since that's pretty neat

PixelyIon commented on app/src/main/cpp/skyline/gpu/command_scheduler.cpp:
```diff
@@ -0,0 +1,53 @@
+
+    void CommandScheduler::SubmitCommandBuffer(const vk::raii::CommandBuffer &commandBuffer, vk::Fence fence) {
+        std::lock_guard lock(gpu.queueMutex);
+        gpu.vkQueue.submit(vk::SubmitInfo{
```
What would I check? It doesn't return anything

bylaws replied:
I didn't know that

PixelyIon commented on app/src/main/cpp/skyline/gpu/command_scheduler.cpp:
```diff
@@ -0,0 +1,53 @@
+namespace skyline::gpu {
+    CommandScheduler::CommandBufferSlot::CommandBufferSlot(vk::raii::Device &device, vk::CommandBuffer commandBuffer, vk::raii::CommandPool &pool) : active(true), device(device), commandBuffer(device, commandBuffer, pool), fence(device, vk::FenceCreateInfo{}), cycle(std::make_shared<FenceCycle>(device, *fence)) {}
+
+    bool CommandScheduler::CommandBufferSlot::AllocateIfFree(CommandScheduler::CommandBufferSlot &slot) {
```
`CommandBufferSlot` is `private` so this isn't possible, I don't want to make it public either

bylaws replied:
Oh I see

PixelyIon commented on app/src/main/cpp/skyline/gpu/fence_cycle.h:
```diff
@@ -0,0 +1,134 @@
+        /**
+         * @brief Wait on a fence cycle till it has been signalled
+         */
+        void Wait() {
```
They're all tiny functions, we generally end up keeping these in headers, the entire class is a pretty minimal wrapper over a Vulkan fence so it doesn't introduce much in the way of compilation overhead (Esp. considering that Vulkan-Hpp is also included)

bylaws replied:
yes but literally everything vulkan includes this so move them

PixelyIon replied:
It's a negligible amount of overhead, it's written almost as an extension to Vulkan-Hpp

PixelyIon commented on app/src/main/cpp/skyline/gpu/fence_cycle.h:
```diff
@@ -0,0 +1,134 @@
+        }
+
+        ~FenceCycle() {
+            Wait();
```
If that happens then the GPU driver will likely throw an error that isn't timeout-related, we should depend on that rather than introduce an arbitrary timeout

PixelyIon commented on app/src/main/cpp/skyline/gpu/fence_cycle.h:
```diff
@@ -0,0 +1,134 @@
+            if (!signalled.test(std::memory_order_consume)) {
+                auto it{dependencies.begin()}, next{std::next(it)};
+                if (it != dependencies.end()) {
+                    while (next != dependencies.end()) {
```
I think `for`
```cpp
for (; next != dependencies.end(); next++) {
    (*it)->next = *next;
    it = next;
}
```
looks worse than `while`
```cpp
while (next != dependencies.end()) {
    (*it)->next = *next;
    it = next;
    next = std::next(next);
}
```
but I don't want to bikeshed so I'll let you have it your way

PixelyIon commented on app/src/main/cpp/skyline/gpu/memory_manager.cpp:
```diff
@@ -0,0 +1,126 @@
+    }
+
+    u8 *Image::data() {
+        if (pointer) [[likely]]
```
It will only not be the case on the first invocation, all further invocations should have the `Image` allocated on the CPU

bylaws replied:
I mean isn't it a forward jump? is it realy needed here

PixelyIon commented on app/src/main/cpp/skyline/gpu/memory_manager.cpp:
```diff
@@ -0,0 +1,126 @@
+            .vulkanApiVersion = GPU::VkApiVersion,
+        };
+        ThrowOnFail(vmaCreateAllocator(&allocatorCreateInfo, &vmaAllocator));
+        // TODO: Use VK_KHR_dedicated_allocation when available (Should be on Adreno GPUs)
```
It'll be done later, I need to implement a Vulkan quirks system where it registers optional features/extensions/other device-specific quirks

PixelyIon commented on app/src/main/cpp/skyline/gpu/memory_manager.cpp:
```diff
@@ -0,0 +1,126 @@
+        VkBuffer buffer;
+        VmaAllocation allocation;
+        VmaAllocationInfo allocationInfo;
+        ThrowOnFail(vmaCreateBuffer(vmaAllocator, &static_cast<const VkBufferCreateInfo &>(bufferCreateInfo), &allocationCreateInfo, &buffer, &allocation, &allocationInfo));
```
Sadly, it cannot as the cast from `vk::ImageCreateInfo` (Vulkan-Hpp type) to `VkImageCreateInfo` (Vulkan C type) is done by defining `vk::ImageCreateInfo::operator VkImageCreateInfo &()` which needs to be casted to using an explicit cast prior to taking a pointer value since the pointer value becomes `vk::ImageCreateInfo*` that cannot be implicitly casted to `VkImageCreateInfo*` as pointer values do not inherit the implicit casts of their underlying types.

PixelyIon commented on app/src/main/cpp/skyline/gpu/memory_manager.h:
```diff
@@ -0,0 +1,93 @@
+        VmaAllocation vmaAllocation;
+        vk::Buffer vkBuffer;
+
+        constexpr StagingBuffer(u8 *pointer, size_t size, VmaAllocator vmaAllocator, vk::Buffer vkBuffer, VmaAllocation vmaAllocation) : vmaAllocator(vmaAllocator), vkBuffer(vkBuffer), vmaAllocation(vmaAllocation), span(pointer, size) {}
```
I wouldn't

PixelyIon commented on app/src/main/cpp/skyline/gpu/native_window.h:
```diff
@@ -0,0 +1,303 @@
+// Copyright © 2021 The Android Open Source Project
+
+#pragma once
+
```
We're not having file comments, they're inconsistent and weird to have because of how blatantly simple file-based categorization generally ends up being like you could easily collect that information in the time.. I didn't realize this was the native window file I agree now but I took a lot of time to write up this comment so I will post it regardless so you have an equivalent amount of time wasted by reading it.

bylaws replied:
😆

PixelyIon commented on app/src/main/cpp/skyline/gpu/presentation_engine.cpp:
```diff
@@ -2,83 +2,288 @@
+        engine->lastChoreographerTime = frameTimeNanos;
+        engine->vsyncEvent->Signal();
+
+        AChoreographer_postFrameCallback(AChoreographer_getInstance(), reinterpret_cast<AChoreographer_frameCallback>(&ChoreographerCallback), engine);
```
We cannot use it as [`AChoreographer_postFrameCallback64`](https://developer.android.com/ndk/reference/group/choreographer#achoreographer_postframecallback64) is only available on API version 29 or later, I would have to have `dlopen` with branching depending on the API version. I could also up the API version I suppose as we already sort of decided this but I didn't go through with it?

bylaws replied:
up the API version then :)

PixelyIon commented on app/src/main/cpp/skyline/gpu/presentation_engine.cpp:
```diff
@@ -2,83 +2,288 @@
+                UpdateSwapchain(swapchainFormat, swapchainExtent);
+
+            if (window->common.magic != AndroidNativeWindowMagic)
+                throw exception("ANativeWindow* has unexpected magic: {} instead of {}", span(&window->common.magic, 1).as_string(true), span<const u8>(reinterpret_cast<const u8 *>(&AndroidNativeWindowMagic), sizeof(u32)).as_string(true));
```
It's not null-terminated rather *null or `sizeof(u32)`* whichever comes first, `libfmt` doesn't handle something having multiple null terminators in a single string and ends up passing it through to the final string which then does an incomplete print

PixelyIon commented on app/src/main/cpp/skyline/gpu/presentation_engine.cpp:
```diff
@@ -2,83 +2,288 @@
+            if (windowScalingMode != NativeWindowScalingMode::ScaleToWindow)
+                window->perform(window, NATIVE_WINDOW_SET_SCALING_MODE, static_cast<i32>(windowScalingMode));
+
+            window->perform(window, NATIVE_WINDOW_ENABLE_FRAME_TIMESTAMPS, true);
```
I forgot to add one, good catch!

PixelyIon committed:
Address CR Comments + Increase `minSdkVersion` to 29 (Android 10)

bylaws commented on .idea/jarRepositories.xml:
```diff
@@ -36,5 +36,10 @@
       <option name="name" value="MavenLocal" />
```
Not doing this?

bylaws replied:
This file still hasn't been deleted

bylaws commented:
> The following items were performed in this PR:
> * Implement [Vulkan](https://www.vulkan.org/) [`Instance`](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkInstance.html) + [`PhysicalDevice`](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPhysicalDevice.html) + [`Device`](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkDevice.html) initialization with [`VK_EXT_debug_report`](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VK_EXT_debug_report.html) support
> * Integrate [`VK_LAYER_KHRONOS_validation`](https://github.com/KhronosGroup/Vulkan-ValidationLayers/blob/master/docs/khronos_validation_layer.md) from Android NDK into debug builds
> * Implement base Vulkan infrastructure (`FenceCycle`, Memory Manager [using [VMA](https://github.com/GPUOpen-LibrariesAndSDKs/VulkanMemoryAllocator)] and Command Scheduler)
> * Rework `GuestTexture` + `Texture` classes with Vulkan infrastructure
> * Implement Vulkan-based Presentation Engine
> * Isolate guest and host presentation for allowing [n-Buffering](https://en.wikipedia.org/wiki/Multiple_buffering) on host
> * Utilize [`AChoreographer`](https://developer.android.com/ndk/reference/group/choreographer) for accurate V-Sync events and adaptive refresh rate updates
> * Dynamic swap-interval management using [`VK_GOOGLE_display_timing`](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VK_GOOGLE_display_timing.html)
> * Implement cropping, transformation and scaling using host [Surfaceflinger](https://source.android.com/devices/graphics/surfaceflinger-windowmanager) for [HWC](https://source.android.com/devices/graphics/hwc) acceleration
> * Rework Performance Statistics to provide a more accurate representation of performance
> * Introduce `Display` settings for fine-grained control over presentation
> * Rework `GraphicBufferProducer` for accuracy and (re)implement the following transactions:
> * * `RequestBuffer`
> * * `SetBufferCount`
> * * `DequeueBuffer`
> * * `DetachBuffer`
> * * `DetachNextBuffer`
> * * `AttachBuffer`
> * * `QueueBuffer`
> * * `CancelBuffer`
> * * `Query`
> * * `Connect`
> * * `Disconnect`
> * * `SetPreallocatedBuffer`
> * Rework Host1X Syncpoint for code quality and performance
> * Fix race condition in Host1X Syncpoints affecting titles like [Sonic Mania](https://www.nintendo.com/games/detail/sonic-mania-switch/)
> * Rework [Display Services (`VI`)](https://switchbrew.org/wiki/Display_services) + [`IHOSBinderDriver`](https://switchbrew.org/wiki/Nvnflinger_services#dispdrv) for accuracy
> * Implement SVC [`UnmapSharedMemory`](https://switchbrew.org/wiki/SVC#UnmapSharedMemory)
> 
> Closes #197 
> Closes #198

I approve

bylaws reviewed (approved):
Assuming you delete jarRepositories

PixelyIon committed:
Move Mbed TLS to submodule + Update Java Libraries

We've moved from using an AAR for Mbed TLS to a submodule as the AAR was packaged manually and used from a local repository which ended up being very hacky and resulted in Linter errors, it could also not be updated with ease as it would need to be repackaged. All of these issues have been solved by moving to a git submodule tied to the official Mbed TLS GitHub repository.

PixelyIon committed:
Implement `IApplicationDisplayService::ConvertScalingMode`

Utilized by アイリス魔法学園

PixelyIon committed:
Fix Scheduler Core Migration Deadlock

Encountered in 不如帰大乱 when `HOS-3` is awoken at the same time as `HOS-0` called `SvcSetThreadCoreMask` resulting in a deadlock where `HOS-0` owns `HOS-3`'s `coreMigrationMutex` while `HOS-3` owns the core mutex with the both of them attempting to lock the other mutex
